test/core/network/executor_test.dart

```diff
@@ -0,0 +1,53 @@
+import 'package:coffeecard/core/errors/failures.dart';
+import 'package:coffeecard/core/network/executor.dart';
+import 'package:dartz/dartz.dart';
+import 'package:flutter_test/flutter_test.dart';
+import 'package:logger/logger.dart';
+import 'package:mockito/annotations.dart';
+
+import '../../response.dart';
+import 'executor_test.mocks.dart';
+
+@GenerateMocks([Logger])
+void main() {
+  late MockLogger logger;
+  late Executor executor;
+
+  setUp(() {
+    logger = MockLogger();
+    executor = Executor(logger);
+  });
+
+  test('should return [ServerFailure] if api call fails', () async {
+    // arrange
+    final tResponse = Response.fromStatusCode(500);
+
+    // act
+    final actual = await executor(() async => tResponse);
+
+    // assert
+    expect(actual, const Left(ServerFailure('')));
```

Comment:
Check the type instead of checking against the concrete value `Left(ServerFailure(''))`

Reply:
no, i want to differentiate between the exception block and the unsuccessful block. 

the empty string is used because 'Returns the response body as a String provided the network call was successful, else this will be null' in ServerFailure.fromResponse:

 on Exception {
      return ServerFailure(response.bodyString);
    }

this is a limitation of the way we create dummy requests